server/src/main/java/org/elasticsearch/gateway/GatewayService.java

```diff
@@ -220,20 +221,22 @@ public ClusterState execute(final ClusterState currentState) {
                 return currentState;
             }
 
-            final ClusterState newState = Function.<ClusterState>identity()
+            return Function.<ClusterState>identity()
                 .andThen(ClusterStateUpdaters::updateRoutingTable)
                 .andThen(ClusterStateUpdaters::removeStateNotRecoveredBlock)
                 .apply(currentState);
-
-            return allocationService.reroute(newState, "state recovered");
         }
 
         @Override
         public void clusterStateProcessed(final ClusterState oldState, final ClusterState newState) {
             logger.info("recovered [{}] indices into cluster_state", newState.metadata().indices().size());
             // reset flag even though state recovery completed, to ensure that if we subsequently become leader again based on a
             // not-recovered state, that we again do another state recovery.
-            resetRecoveredFlags();
+            if (oldState != newState) {
```

Comment:
I am a bit hesitant to this. I think that if the states are identical, we no longer have the block either. And we *should* only submit the task if we have not recovered yet and only once through the booleans in this class.

I wonder if we should just assert that they are different instead? And then always call reroute since that seems most defensive anyway.

I may have missed the situation making this necessary though, hope you can educate me then.